extension/headerssetter/extension_test.go

```diff
@@ -0,0 +1,228 @@
+		t.Run("round_tripper", func(t *testing.T) {
+			ext, err := newHeadersSetterExtension(tt.cfg)
+			assert.NoError(t, err)
+			assert.NotNil(t, ext)
+
+			roundTripper, err := ext.RoundTripper(mrt)
+			assert.NoError(t, err)
+			assert.NotNil(t, roundTripper)
+
+			ctx := client.NewContext(
+				context.Background(),
+				client.Info{
+					Metadata: tt.metadata,
+				},
+			)
+			req, err := http.NewRequestWithContext(ctx, "GET", "", nil)
+			assert.NoError(t, err)
+			assert.NotNil(t, req)
+
+			resp, err := roundTripper.RoundTrip(req)
+			assert.NoError(t, err)
+			assert.NotNil(t, resp)
+			for _, header := range tt.cfg.HeadersConfig {
+				assert.Equal(
+					t,
+					resp.Header.Get(*header.Key),
+					tt.expectedHeaders[*header.Key],
+				)
+			}
+		})
+	}
+}
+
+func TestPerRPCCredentials(t *testing.T) {
+	for _, tt := range tests {
+		t.Run("", func(t *testing.T) {
+			ext, err := newHeadersSetterExtension(tt.cfg)
+			assert.NoError(t, err)
+			assert.NotNil(t, ext)
+
+			perRPC, err := ext.PerRPCCredentials()
+			assert.NoError(t, err)
+			assert.NotNil(t, perRPC)
+
+			ctx := client.NewContext(
+				context.Background(),
+				client.Info{Metadata: tt.metadata},
+			)
+
+			metadata, err := perRPC.GetRequestMetadata(ctx)
+			assert.NoError(t, err)
+			assert.NotNil(t, metadata)
+			for _, header := range tt.cfg.HeadersConfig {
+				assert.Equal(
+					t,
+					metadata[*header.Key],
+					tt.expectedHeaders[*header.Key],
+				)
+			}
+		})
+	}
+}
+
+var (
+	mrt           = &mockRoundTripper{}
+	header        = "header_name"
+	anotherHeader = "another_header_name"
+	tests         = []struct {
+		cfg             *Config
+		metadata        client.Metadata
+		expectedHeaders map[string]string
+	}{
+		{
+			cfg: &Config{
+				HeadersConfig: []HeaderConfig{
+					{
+						Key:         &header,
+						FromContext: stringp("tenant"),
+					},
+				},
+			},
+			metadata: client.NewMetadata(
+				map[string][]string{"tenant": {"context value"}},
+			),
+			expectedHeaders: map[string]string{
+				"header_name": "context value",
+			},
+		},
+		{
+			cfg: &Config{
+				HeadersConfig: []HeaderConfig{
+					{
+						Key:   &header,
+						Value: stringp("config value"),
+					},
+				},
+			},
+			expectedHeaders: map[string]string{
+				"header_name": "config value",
+			},
+		},
+		{
+			cfg: &Config{
+				HeadersConfig: []HeaderConfig{
+					{
+						Key:         &header,
+						FromContext: stringp("tenant"),
+					},
+					{
+						Key:         &anotherHeader,
+						FromContext: stringp("tenant"),
+					},
+				},
+			},
+			metadata: client.NewMetadata(
+				map[string][]string{"tenant": {"acme"}},
+			),
+			expectedHeaders: map[string]string{
+				"header_name":         "acme",
+				"another_header_name": "acme",
+			},
+		},
+		{
+			cfg: &Config{
+				HeadersConfig: []HeaderConfig{
+					{
+						Key:         &header,
+						FromContext: stringp(""),
+					},
+				},
+			},
+			expectedHeaders: map[string]string{
+				"header_name": "",
```

Comment:
I guess this answers the question I had earlier about empty FromContext values.

Reply:
The semantic is following:
```
from_context: "key" -> if key is missing from context then header value set to ""
from_context: "key" -> if key from context is "" then header value set to ""
value: "" -> header value set to ""
```

I was thinking of not setting headers at all if their value is "" but then decided to make arguably more explicit. Also, "" is a valid header value.